velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -458,65 +467,102 @@ struct DecimalDivideFunction {
       uint8_t aScale,
       uint8_t bPrecision,
       uint8_t bScale) {
-    auto scale = std::max(6, aScale + bPrecision + 1);
-    auto precision = aPrecision - aScale + bScale + scale;
-    return DecimalUtil::adjustPrecisionScale(precision, scale);
+    if constexpr (allowPrecisionLoss) {
+      auto scale = std::max(6, aScale + bPrecision + 1);
+      auto precision = aPrecision - aScale + bScale + scale;
+      return DecimalUtil::adjustPrecisionScale(precision, scale);
+    } else {
+      auto wholeDigits = std::min(38, aPrecision - aScale + bScale);
+      auto fractionDigits = std::min(38, std::max(6, aScale + bPrecision + 1));
+      auto diff = (wholeDigits + fractionDigits) - 38;
+      if (diff > 0) {
+        fractionDigits -= diff / 2 + 1;
+        wholeDigits = 38 - fractionDigits;
+      }
+      return DecimalUtil::bounded(wholeDigits + fractionDigits, fractionDigits);
+    }
   }
 
   uint8_t aRescale_;
   uint8_t rPrecision_;
 };
 
-template <template <class> typename Func>
+template <template <class, bool> typename T, bool allowPrecisionLoss>
+using ParameterBinder = TempWrapper<T<exec::VectorExec, allowPrecisionLoss>>;
+
+template <typename Func, typename TReturn, typename... TArgs>
+bool registerFunction(
+    const std::vector<std::string>& aliases = {},
+    const std::vector<exec::SignatureVariable>& constraints = {},
+    bool overwrite = true) {
```

Comment:
This function seems to be the same as an existing function. Can we reuse it?

https://github.com/facebookincubator/velox/blob/e13c591f49b04ddbfc79f03335360f29ed574e3a/velox/functions/Registerer.h#L63

Reply:
No, template bool is different, and `using funcClass = typename Func::template udf<exec::VectorExec>;` is different, so I add new binder to support template function with variable bool. 
```
template <template <class, bool> typename T, bool allowPrecisionLoss>
using ParameterBinder = TempWrapper<T<exec::VectorExec, allowPrecisionLoss>>;
```

Reply:
I wonder if we should move it to [velox/velox/functions/Registerer.h](https://github.com/facebookincubator/velox/blob/e13c591f49b04ddbfc79f03335360f29ed574e3a/velox/functions/Registerer.h#L63) in a separate PR?
It could allow function with constraints accept template.

Reply:
It seems compared with the below function, the new one added 'constraints'. It this the only difference? Thanks.

https://github.com/facebookincubator/velox/blob/e13c591f49b04ddbfc79f03335360f29ed574e3a/velox/functions/Registerer.h#L45-L47

Reply:
Yes. I wonder if we can consolidate these APIs.

Reply:
I agree. Perhaps we can change the API in a separate PR.

Reply:
Do we need to only keep this one or keep both functions but the latter changes from `Func<exec::VectorExec>` to `typename Func::template udf<exec::VectorExec>`, for the backward compatibility, I would prefer keep both. The first could revoke the latter with the argument `constraints = {}`.
```
bool registerFunction(
    const std::vector<std::string>& aliases = {},
    const std::vector<exec::SignatureVariable>& constraints = {},
    bool overwrite = true) {
using funcClass = typename Func::template udf<exec::VectorExec>;
}
```
As the comment indicates, the wrapper for udf struct will be removed finally, so can I do that? @mbasmanova 
CC @rui-mo 

> // New registration function; mostly a copy from the function above, but taking
> // the inner "udf" struct directly, instead of the wrapper. We can keep both for
> // a while to maintain backwards compatibility, but the idea is to remove the
> // one above eventually.